Add guidance on API documentation (#4160)
kubernetes-sigs/gateway-api: site/content/en/contributing/devguide.md

such as on the Gateway API documentation website, or via `kubectl explain`.

Additionally, it is worth noting that API documentation affects the size of generated CRD manifests,
which directly impacts the Kubernetes resource size, 
which is in turn limited by etcd maximum value size.  The situation is exacerbated by the `last-applied-configuration` 
annotation from client-side apply, which roughly doubles the size of the Kubernetes resource.

There are two kind of API documentation: 

* User facing - MUST define how a user should be consuming an API and its fields, in a concise way.
* Developer facing - MUST define how a controller should implement an API and its fields. 

### User facing Documentation

The API documentation, when meaningful, helps users of it understand and create configuration
in a way that ensures that their Gateway API implementation does what they want it to do.

A good API documentation should cover:

* What is the main feature of the API and Field - Eg.: "`foo` allows configuring how a
a header should be forwarded to backends"
* What is the support level of the field - Eg.: "Support: Core/Extended/Implementation Specific"

> * Caveats of that field - Eg.: "Setting `goo` field can have conflicts with `bar` field, and in this

Reviewer: ```suggestion
* Caveats of that field - Eg.: "Setting `foo` field can have conflicts with `bar` field, and in this
```

Author: updated on my last commit

behavior of it, and should answer questions such as the following:

* How should that field be reconciled?
* What conditions should be set during the reconciliation? 
* What should be validated during the reconciliation of that field?

In this case, as the API documentation serves as a guide for implementers on how 
their implementations should behave, it is very important to be as verbose as
required to avoid any ambiguity. This information is used also to define expected
conformance behavior, and it can point to existing GEPs so a developer looking 
at it can know where to look for more references on what the expected
behavior of this field is and why.

Continuing with the `Listeners` field as an example, it shows how API documentation can provide guidance for situations 
such as the following:

* Two listeners have different protocols but the same hostname. Should this be a conflict?
* A listener of type `XXX` sets the field `TLS`. Is this a problem? How to expose this to 
the user?

Because these details don't matter for a user, they should be hidden from the CRD/OpenAPI
generation and also from the website API Reference.

> This can be achieved putting these information between the tags 

Reviewer 2: ```suggestion
This can be achieved by putting the information between the tags 
```

Author: fixed

`<gateway:util:excludeFromCRD></gateway:util:excludeFromCRD>` and preferably 

> contain a callout that those are a Note for implementers:

Reviewer 2: ```suggestion
contain a callout that this is a Note for implementers:
```

Author: done
